feat(core): add notification feedback with angular-ui-notification

sujeethk opened the pull request:
feat(core): add notification feedback with angular-ui-notification

Added visual notification for user/article updates
angular-ui-notification config added to core client config
Notification idea from #369

sujeethk commented on modules/articles/client/controllers/admin/article.client.controller.js:
```diff
@@ -20,7 +20,10 @@
     function remove() {
       if ($window.confirm('Are you sure you want to delete?')) {
-        vm.article.$remove($state.go('admin.articles.list'));
+        vm.article.$remove(function() {
```
Fix for #1274

sujeethk commented on modules/core/client/app/config.js:
```diff
@@ -20,4 +20,16 @@
   }
+
+  // Toastr Config Global settings. For more customization check https://github.com/Foxandxss/angular-toastr
+  angular.module('toastr').config(function(toastrConfig) {
```
Toastr config

sujeethk commented on modules/core/server/controllers/core.server.controller.js:
```diff
@@ -14,6 +14,7 @@ exports.renderIndex = function (req, res) {
       displayName: validator.escape(req.user.displayName),
       provider: validator.escape(req.user.provider),
       username: validator.escape(req.user.username),
+      _id: req.user._id,
```
Fix #1531

mleanos replied:
This would indeed fix the issue. However, we purposely removed the `_id` field from the client-side User object for reasons of security. I don't see an easy way to handle this on the client-side.

One thing for sure is that we need to have a check on the server-side, to prevent a user from deleting their own account. This can easily be done [here](https://github.com/meanjs/mean/blob/master/modules/users/server/controllers/admin.server.controller.js#L46) by adding something like:
```javascript
if (user._id === req.user._id) {
  return res.status(400).send({
      message: 'You cannot delete yourself!'
    });
}
```

Additionally, we can do something similar to [this](https://github.com/meanjs/mean/blob/master/modules/articles/server/controllers/articles.server.controller.js#L36-L38). Which we would then have a field attached to the User that is being "edited" that specifies whether or not the current active User is the same as the one being edited. This field could be named something like `isCurrentUserSelf` (or something more clever), and this field would NOT be persisted to the database.

sujeethk replied:
I wanted to keep the change simple. How about this?

`<a class="btn btn-primary" ng-click="vm.remove()" ng-if="vm.user.username !== vm.authentication.user.username">`

This should fix the issue + no extra code change needed except changing the _id to username.

mleanos replied:
https://github.com/meanjs/mean/pull/1535

mleanos commented:
I'm not 100% sold on the need for adding a new library to the framework, just to have a toastr feature. However, this implementation is pretty light-weight & if others are find with adding this, then I'm OK with it.

@meanjs/contributors @Wuntenn @hyperreality @simison WDYT?

simison commented:
Toastr seems like a great library otherwise but it requires jQuery. :-(

In general some kind of notifications seem like a basic enough feature for this boilerplate.

simison commented:
It seems like this PR is trying to fix many issues and should be divided to 3 different PRs, or did I understand something wrong?

mleanos commented:
@simison I agree about separating this out into multiple PR's.

So the `angular-toastr` library uses JQuery? That is a bummer.

I'd rather see a more robust notification system with SocketIO. Although with my app I use a Notification service that uses SocketIO, which pops toasts. It was a big implementation, and I'm not even sure I handled it all correctly. 

Having some sort of notification would be great though.

simison commented:
Similar (size/features/maintained) to Toastr but angular-only: https://github.com/alexcrack/angular-ui-notification

Wuntenn commented:
I finding toasts really useful in MD and @simison's last link (angular-ui-notification) doesn't need JQuery which is a plus. They make all kinds of errors more consistent across the site and save loads of unnecessary headache wondering how to present stuff - I'd say it's a good call.

I removed UIB from my project and replaced it with Angular Material Design a few months back. There's quite a lot of tools that are directly interchangeable with UIB's own and toasts are one of the additional features that I didn't realise how good they were until I started using them.

I'm not sure if or when a move to MD is on the cards for MeanJS (I'd be happy to put together a PR) It's worth looking into. You also get a load of ARIA support and notifications in the console if you don't meet a requirement (usually this means filling in some element attribute). This means you get to create more inclusive experiences as you go without tons of upfront research into ARIA.

sujeethk commented:
Split the PR, this one takes care of the notification. I can open seperate PR for the fixes. 

Took your advice and switched library to angular-ui-notification and kept it simple

mleanos commented:
> @mleanos What's the benefit of having socket.io notifications over just Angular controller handling them?

@simison I have notifications that come from the server, not just the client-side. I opted for a client-side service that "pops" toasts (which uses a toastr-like angular service) that can be triggered from my Socket.IO event listeners, or by merely calling `Notifications.pop()` that accepts additional options (e.g. data, type, actions). So while neither my Socket.IO implementation nor my Toaster service, relies on the other, my notifications use both where needed.

I guess after hearing myself explain this, it's more a specific requirement of my application. For this framework, it makes sense to just keep it simple as @sujeethk has done.

mleanos commented:
👍 on using angular-ui-notification. I'll probably end up switching my application to use it.

mleanos commented on modules/articles/client/controllers/admin/article.client.controller.js:
```diff
@@ -21,6 +21,7 @@
     function remove() {
       if ($window.confirm('Are you sure you want to delete?')) {
         vm.article.$remove($state.go('admin.articles.list'));
+        Notification.success('Article deleted successfully!');
```
@sujeethk After merging #1534 this is creating a merge conflict. It's probably best to put this in the `$remove` callback. Can you rebase, and this line?

mleanos commented on modules/users/client/controllers/admin/user.client.controller.js:
```diff
@@ -42,8 +44,10 @@
+        Notification.success('User updated successfully!');
       }, function (errorResponse) {
         vm.error = errorResponse.data.message;
+        Notification.error('User update error!');
```
@sujeethk Do we still need to set `vm.error` in these cases? Or would it be better to add the response error as the body of the toast, and keep "User update error!" as the title?

sujeethk replied:
@mleanos I had it set that way on the angular-toastr, looks like i kept it generic on this one. It is good idea to have the error on the toast and add title.

Removing vm.error would fail some tests that are checking it. Didn't find any harm in having it until we decided to use these notifications.

I will update the error message with title and post a screenshot. Will update tests, although frankly I have no idea how the test scripts are setup(never looked other than running them).

sujeethk commented:
Let me know if I over did it. I believe all form submission feedback is now using the new notifications.

- Added 3 User e2e tests that check the new popup content
- vm.error or vm.success is left as is. Since they are part of the unit testing for controllers. Removing them will prevent a few error case scenarios from being tested
- But removed the error/success blocks from the view files and removed a couple of success/invalid pages in the reset password flow and related routes

I do not have an email service account to do some manual testing around the reset flows. Can someone with access please let me know if those flows are good.

Some screenshots:
![success](https://cloud.githubusercontent.com/assets/19917701/19011910/e8dd5ad4-8772-11e6-994d-76d224553296.png)
![error](https://cloud.githubusercontent.com/assets/19917701/19011911/e8dd7dde-8772-11e6-80c5-6cee75d8b88d.png)

jloveland commented:
Can you post a picture of this working on mobile?

jloveland commented:
seems to be working nicely on mobile for the login screen:
![image1](https://cloud.githubusercontent.com/assets/3456708/19014706/392ab410-87c2-11e6-8ce5-638ac4a8a70e.PNG)

mleanos commented on modules/users/client/config/users.client.routes.js:
```diff
@@ -101,20 +101,6 @@
         url: '/reset',
         template: '<ui-view/>'
       })
-      .state('password.reset.invalid', {
```
The removal of this route is causing a 400 response on the server, with [this redirect](https://github.com/meanjs/mean/blob/master/modules/users/server/controllers/users/users.password.server.controller.js#L115). 

This keeps the User on the Reset Password view if they click on an invalid reset link (e.g. the token is expired). We should keep this route since the User should be taken away from the Reset Password view, in the case of an invalid token; they wouldn't be able to do anything there anyways.

I see there are some good reasons to get rid of the *Success* route, but by keeping it we provide more flexibility for our users. And we could always redirect somewhere else in the case of an *Invalid* token, but I don't see a need to change this flow now.

IMO, we should keep the flow with these two routes. There can probably be some improvements to the Reset Password feature, but I think it's a bit out of scope for this PR.

sujeethk replied:
@mleanos really good feedback, appreciate it! 

I redirected the success to home with a toast. Looks like I missed the error route. Honestly I haven't see these flows in action as I don't have a MSP. Can someone recommend some simple free MSP so I can test these flows?

I simply removed those routes and deleted those two view files to reduce footprint. In those terms I could have the invalid scenario redirect to forgot password screen with an invalid/expired token toast so end-user can request another.

Or the opposite being what you recommend to keep the routes and views in case users need them.

I am neutral to both scenarios. Which one do you prefer?

mleanos replied:
I'd prefer to keep the routes/views. We could always open up a discussion on how to change this flow, but I think having that discussion would distract us from the purpose of this PR.

I'm open to the discussion though, if others feel this should be addressed now.

mleanos commented on modules/users/tests/client/users.client.routes.tests.js:
```diff
@@ -257,44 +257,6 @@
         });
       });
 
-      describe('Invalid Route', function () {
```
Keep these tests if we end up keeping the client-side routes; based on my comment above.

mleanos commented on modules/users/tests/client/password.client.controller.tests.js:
```diff
@@ -189,7 +189,7 @@
           });
 
           it('should redirect to password reset success view', function() {
-            expect($location.path).toHaveBeenCalledWith('/password/reset/success');
```
Related to my comment above.

mleanos commented on modules/users/client/controllers/settings/change-password.client.controller.js:
```diff
@@ -33,11 +33,13 @@
       // If successful show success message and clear form
       $scope.$broadcast('show-errors-reset', 'vm.passwordForm');
       vm.success = true;
```
Since these aren't used by the view any longer, we should remove them from the controller.

@sujeethk I know you mentioned having issues with the client-side tests that relied on these `vm` fields, but they you should be able to resolve them. Let me know if you have trouble, and I can take a look.

sujeethk replied:
@mleanos the couple of lines you mentioned above have been removed with no issues with tests. But there are a few tests on other controllers that check for the status of vm.success and vm.error.

```javascript
it('should set success to response message', function() {
  expect(scope.vm.success).toBe(successMessage);
});
```

Do you want me to remove all of those tests? If not can you please let me know how I can check for a DOM element from the test script?

mleanos replied:
Is it possible to modify this test (and others like it) to check that the toast was opened?

simison replied:
Indeed, otherwise nothing is testing if the notification actually worked.

sujeethk replied:
@mleanos @simison I added about 3 tests on user e2e that is checking for the notification and it's contents.

But I am stuck on these client controller tests. If we remove the vm.error from the controller and intend to continue unit testing the controllers for error we would need to check for the notification. The notification is DOM element which I don't have access to without using jasmine jquery. I know we can access DOM in protractor  e2e tests. But any guidance on how I can check for the notification on the controller tests?

mleanos replied:
Im not at a computer right now so i cant give examples, or links..

In some of our Karma tests, you'll see we have a Spy on some of the angular controller services. You can add some in these tests that ensure the Notification service was called; I forget the syntax. If a simple spy doesnt work, you can look into Sinon's Stub feature.

sujeethk replied:
Sure will take a look. Thank you.

mleanos replied:
https://github.com/meanjs/mean/blob/master/modules/core/tests/client/interceptors/auth-interceptor.client.tests.js#L26
https://github.com/meanjs/mean/blob/master/modules/core/tests/client/interceptors/auth-interceptor.client.tests.js#L49

Something similar can be done. The first links shows how to setup the Spy to watch the specific service method. You can setup a Spy for each Notification method that you expect to be called; in this case, `Notification.error()` & `Notification.success()`.

sujeethk replied:
Removed all vm.success and vm.error and replaced any existing tests for error or success to check for Notification called. All tests passed on my local. Please review carefully :), first time using Jasmine spies.

mleanos replied:
Ok. Thank you. I'll pull this branch down again, and test locally.

mleanos commented on modules/users/client/controllers/admin/user.client.controller.js:
```diff
@@ -42,8 +44,9 @@
         $state.go('admin.user', {
           userId: user._id
         });
+        Notification.success({ message: '<i class="glyphicon glyphicon-ok"></i> User deleted successfully!' });
```
Should be "User saved successfully!"

mleanos commented on modules/users/client/config/users.client.routes.js:
```diff
@@ -115,6 +101,20 @@
           pageTitle: 'Password reset success'
         }
       })
+      .state('password.forgot', {
```
Not critical, but can these state's be added back to where they were before these changes. It's causing an unnecessary diff. This can be misleading when looking back at the history.

mleanos commented on modules/users/client/controllers/password.client.controller.js:
```diff
@@ -55,13 +53,13 @@
       // Show user success message and clear form
       vm.credentials = null;
-      vm.success = response.message;
+      Notification.success({ message: response.message, title: 'Password request successful!' });
```
Very nit-picky, but can be "Password Reset email successfully sent" or something more user friendly.

sujeethk replied:
@mleanos That text is just the title of the error notification. The actual message from the server would be the message like(which is already in the server controllers):
`An email has been sent to the provided email with further instructions.`
`No account with that username has been found`
So the title would need to be generic to what the request is for right?

mleanos replied:
Sure. I understand. Just thought that the title wasn't very clear, or descriptive, for a user. Feel free to leave it as is.

sujeethk replied:
![image](https://cloud.githubusercontent.com/assets/19917701/19058411/12725a78-89a5-11e6-8512-67268773ace2.png)

@mleanos Did you mean a title like this? I thought the title and the message were decoupled as the email was not attempted to be sent, and so kept the title generic. But if this image makes more sense let me know and I can update the text.

sujeethk replied:
On second thoughts, I actually get your point. The title is misleading. User is not requesting for a password. He is only requesting for a password reset email. I will add a couple of words and commit.

mleanos commented on modules/users/client/controllers/password.client.controller.js:
```diff
@@ -55,13 +53,13 @@
       // Show user error message and clear form
       vm.credentials = null;
-      vm.error = response.data.message;
+      Notification.error({ message: response.data.message, title: '<i class="glyphicon glyphicon-remove"></i> Password request failed!', delay: 4000 });
```
Similar to my nitty comment above.

mleanos commented:
@sujeethk Just a couple of minor changes/feedback, and this should be ready.

sujeethk commented:
@mleanos On a side note, I have been thinking to add an error notification into the `auth-interceptor.client.service` for error code `-1`.

This will show an error notification if the user lost his connection or the request timed-out or God forbid the server goes down. In these cases the client controllers will not be able to show their notifications(or vm.error in current master) as there is no response. Even if it is called res.data.message is not available, and there would be an angular error posted to the console(which probably the client isnt watching). So end user experience is that he clicks submit and waits, nothing happens.

Do you think it is good to have? If so, do you think I can add it in this PR or does it need a seperate PR?

mleanos commented:
@sujeethk Let's leave the auth-interceptor Notifications for a separate PR & discussion. It would be better to wrap this PR up, to get us closer to the 0.5.0 release.

mleanos commented:
@meanjs/contributors @simison @jloveland LGTY?

codydaig commented:
My only question: In most controllers, I see `Notification.sucess()` after `$state.go()`. Will that ever display and is that just excess code?

mleanos commented:
@codydaig The Notification method still gets called & will display, after we transition to the next state.

Since the Notification is applied to the root-scope, the toast is independent of what current state we're on; the toast ends up showing up during our state transition, and remains after the new state is loaded.

codydaig commented:
@mleanos Thanks for the clarification.

mleanos commented:
@sujeethk After merging #1499 there are merge conflicts with this PR. They should be pretty minor. Can you resolve them, and then we can merge this. I apologize for the inconvenience.

sujeethk commented:
@mleanos No worries. Merged and committed.

mleanos commented:
@sujeethk Couple more conflicts. If you can rebase one more time, I'll merge right away.

sujeethk commented:
@mleanos I am currently our of station on vacation. Will rebase and merge Monday night. Apologies for the delay

sujeethk commented:
@mleanos Rebased

mleanos commented:
@sujeethk Thank you! Merging now.

@simison Thanks for your feedback as well.
